src/profile-logic/import/dhat.js

```diff
@@ -0,0 +1,324 @@
+    let funcIndex = funcKeyToFuncIndex.get(funcKey);
+    if (funcIndex === undefined) {
+      funcTable.name.push(stringTable.indexForString(funcName));
+      funcTable.isJS.push(false);
+      funcTable.relevantForJS.push(false);
+      funcTable.resource.push(-1);
+      funcTable.fileName.push(stringTable.indexForString(fileName));
+      funcTable.lineNumber.push(line);
+      funcTable.columnNumber.push(column);
+      funcIndex = funcTable.length++;
+      funcKeyToFuncIndex.set(funcKey, funcIndex);
+    }
+
+    frameTable.address.push(address);
+    frameTable.line.push(line);
+    frameTable.column.push(column);
+    frameTable.category.push(otherCategory);
+    frameTable.subcategory.push(otherSubCategory);
+    frameTable.innerWindowID.push(null);
+    frameTable.implementation.push(null);
+    frameTable.func.push(funcIndex);
+    frameTable.length++;
+  }
+
+  const totalBytes: Bytes[] = [];
+  const maximumBytes: Bytes[] = [];
+  const bytesAtGmax: Bytes[] = [];
+  const endBytes: Bytes[] = [];
+
+  for (const pp of dhat.pps) {
+    // Never reset the stackIndex, as it as stack indexes always growing taller.
+    let stackIndex = -1;
+    let prefix = null;
+
+    // Go from root to tip on the backtrace.
+    for (let i = pp.fs.length - 1; i >= 0; i--) {
+      // The dhat frame indexes matches the process profile frame index.
+      const frameIndex = pp.fs[i];
+      const funcIndex = frameTable.func[frameIndex];
+
+      // Start searching for a stack index.
+      for (
+        // Increase the stackIndex by one. For the first run, this will end up
+        // being a stackIndex of 0. For the consecutive runs, it will be the next
+        // stack index.
+        stackIndex++;
+        stackIndex < stackTable.length;
+        stackIndex++
+      ) {
+        const nextFrameIndex = stackTable.frame[stackIndex];
+        if (
+          frameTable.func[nextFrameIndex] === funcIndex &&
+          stackTable.prefix[stackIndex] === prefix
+        ) {
+          break;
+        }
+      }
+
+      if (stackIndex === stackTable.length) {
+        // No stack index was found, add on a new one.
+        stackTable.frame.push(frameIndex);
+        stackTable.category.push(otherCategory);
+        stackTable.category.push(otherSubCategory);
+        stackTable.prefix.push(prefix);
+        // The stack index already points to this spot.
+        stackTable.length++;
+      }
+
+      prefix = stackIndex;
+    }
+
+    // Skip pushing onto the allocation weights, as each byte type will be added
+    // as a separate thread.
+    totalBytes.push(pp.tb);
+    maximumBytes.push(pp.mb);
+    bytesAtGmax.push(pp.gb);
+    endBytes.push(pp.eb);
+
+    allocations.time.push(0);
+    allocations.stack.push(stackIndex);
+    allocations.length++;
+  }
+
+  const totalBytesThread = { ...thread };
+  const maximumBytesThread = { ...thread };
+  const bytesAtGmaxThread = { ...thread };
+  const endBytesThread = { ...thread };
```

Comment:
I think the codebase is pretty good about not mutating things, there's a slight risk, but not sure deep clone is worth it here, due to the complexity of an implementation. If you do want this, I would suggest writing a util called `cloneThread`, that way it can be maintained outside of the importer.

Reply:
I think we mutate in a few locations:
* when we upgrade profiles. And that's where my question above makes more sense too :-)
* when we serialize the profile (when we upload for example) : it would be good to double-check this use case will work as expected.

Reply:
I found a solution for avoiding sharing data structures that I felt comfortable with, see the latest commit.